src/Lecture2.hs

```diff
@@ -164,7 +180,56 @@ data Knight = Knight
     , knightEndurance :: Int
     }
 
-dragonFight = error "TODO"
+data Treasure = Treasure String
+   deriving (Show)
+
+data Chest = Chest
+    { gold          :: Int
+    , maybeTreasure :: Maybe Treasure
+    }
+
+data DragonType = Red | Block | Green
+
+data Dragon = Dragon
+    { dragonType      :: DragonType
+    , dragonHealth    :: Int
+    , dragonFirePower :: Int
+    , dragonStomachs  :: Chest
+    }
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There're two problems with this particular data type definition:

1. It's possible to have a treasure inside green dragon:

```haskell
myDragon = Dragon
    { dragonType      = Green
    , dragonHealth    = 150
    , dragonFirePower = 70
    , dragonStomachs  = Chest 10 (Just (Treasure "some treasure"))
    }
```

2. It's possible to not have a treasure inside non-green dragons:

```haskell
myDragon = Dragon
    { dragonType      = Red
    , dragonHealth    = 150
    , dragonFirePower = 70
    , dragonStomachs  = Chest 10 (Just (Treasure "some treasure"))
    }
    }
```

Reply:
Let me update this with a simpler implementation.